node/network/availability-recovery/src/lib.rs

```diff
@@ -0,0 +1,724 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Availability Recovery Subsystem of Polkadot.
+
+#![warn(missing_docs)]
+
+use std::collections::HashMap;
+
+use futures::prelude::*;
+use futures::stream::FuturesUnordered;
+use futures::channel::{oneshot, mpsc};
+use lru::LruCache;
+
+use polkadot_primitives::v1::{
+	AvailableData, CandidateReceipt, CandidateHash,
+	Hash, ErasureChunk, ValidatorId, ValidatorIndex,
+	SessionInfo, SessionIndex, BlakeTwo256, HashT,
+};
+use polkadot_subsystem::{
+	SubsystemContext, SubsystemResult, SubsystemError, Subsystem, SpawnedSubsystem, FromOverseer,
+	OverseerSignal, ActiveLeavesUpdate,
+	errors::RecoveryError,
+	messages::{
+		AvailabilityStoreMessage, AvailabilityRecoveryMessage, AllMessages, NetworkBridgeMessage,
+	},
+};
+use polkadot_node_network_protocol::{
+	v1 as protocol_v1, NetworkBridgeEvent, PeerId, ReputationChange as Rep, RequestId,
+};
+use polkadot_node_subsystem_util::{
+	validator_discovery,
+	request_session_index_for_child_ctx,
+	request_session_info_ctx,
+};
+use polkadot_erasure_coding::branch_hash;
+
+mod error;
+
+#[cfg(test)]
+mod tests;
+
+const LOG_TARGET: &str = "availability_recovery";
+
+const COST_MERKLE_PROOF_INVALID: Rep = Rep::new(-100, "Merkle proof was invalid");
+const COST_UNEXPECTED_CHUNK: Rep = Rep::new(-100, "Peer has sent an unexpected chunk");
+
+// How many parallel requests interaction should have going at once.
+const N_PARALLEL: usize = 50;
+
+// Size of the LRU cache where we keep recovered data.
+const LRU_SIZE: usize = 16;
+
+/// Compute the threshold from the number of validators.
+///
+/// It should be f + 1, where n = 3f + k, where k in {1, 2, 3}, and n is the number of validators.
+const fn recovery_threshold(n_validators: usize) -> usize {
+	let k = n_validators % 4;
+	let f = (n_validators - k) / 3;
+	f + 1
+}
+
+/// The Availability Recovery Subsystem.
+pub struct AvailabilityRecoverySubsystem;
+
+type ChunkResponse = Result<(PeerId, ErasureChunk), RecoveryError>;
+
+/// Data we keep around for every chunk that we are awaiting.
+struct AwaitedChunk {
```

Comment:
The guide also has an `issued_at` field. Either the guide or this struct should be updated, so they're in sync.

Reply:
I think I wanted `issued_at` to handle request timeouts. Looking over the guide text now, I don't see where that is referenced. I suppose it was not included here because it wasn't used.

I suggest that we alter the guide and the implementation here to keep `issued_at`, but to prune out any entries from both `discovering_validators` and `live_chunk_requests` which have been around longer than some constant `REQUEST_TIMEOUT`. We'll have to poll periodically in order to do this garbage collection.

Dropping them from the state should lead to the interaction receiving a notification from `requesting_chunks` of `oneshot::Canceled`. Within the interaction, we should roughly just ignore such failures as the loop will request from other validators as a replacement.

Reply:
Now I see that there's a `CHUNK_REQUEST_TIMEOUT`. Still need to see how this is used to clean up the `state`.

Reply:
As a rule, we could just move on to another chunk whenever one looks sticky, which amounts to some balancing act between waiting for timeouts and preemptively adding more requests.

Reply:
@burdges in practice that would mean having two timeouts a 'sticky' timeout and a 'hard-fail' timeout. seems worth exploring but not for the initial PR i'd say